src/Compilers/CSharp/Portable/Declarations/DeclarationTreeBuilder.cs

```diff
@@ -898,6 +966,54 @@ static bool checkPropertyOrFieldMemberForRequiredModifier(Syntax.InternalSyntax.
             }
         }
 
+        private ImmutableSegmentedHashSet<string> GetOrComputeMemberNames<TData>(
+            SyntaxNode parent,
+            Action<HashSet<string>, TData> addMemberNames,
+            TData data)
+        {
+            // Try to obtain the prior member names computed for the corresponding type decl (in lexicographic order)
+            // from the prior version of this tree.
+            var previousMemberNames = _currentTypeIndex < _previousMemberNames.Count
+                ? _previousMemberNames[_currentTypeIndex]
+                : ImmutableSegmentedHashSet<string>.Empty;
+
+            _currentTypeIndex++;
+
+            // Lookup in the cache first.
+            var greenNode = parent.Green;
+            if (!s_nodeToMemberNames.TryGetValue(greenNode, out var memberNames))
+            {
+                // If not there, make a fresh set, and add all the member names to it.
+                var memberNamesBuilder = PooledHashSet<string>.GetInstance();
+                addMemberNames(memberNamesBuilder, data);
+
+                // If the members names are the same as the prior computed ones, then use that instead.
+                var result = previousMemberNames.Count == memberNamesBuilder.Count && previousMemberNames.SetEquals(memberNamesBuilder)
+                    ? previousMemberNames
+                    : ImmutableSegmentedHashSet.CreateRange(memberNamesBuilder);
+                memberNamesBuilder.Free();
+
+                // Don't bother caching when there are no member names (common for most compilation units).
+                if (result.Count == 0)
+                    return result;
+
+                // Store the names in the cache to be found in the next compilation update.
+                memberNames = new StrongBox<ImmutableSegmentedHashSet<string>>(result);
+
+#if NET
+                s_nodeToMemberNames.AddOrUpdate(greenNode, memberNames);
+#else
+                lock (s_nodeToMemberNames)
```

Comment:
>lock (s_nodeToMemberNames) [](http://example.com/codeflow?start=16&length=26)

I am not sure if there is any real value from using a lock here. Each operation is thread safe. And we probably do not want to overwrite the value if we lost the race, instead we should share the value with the winner. So, I suggest:
- removing the conditional compilation
- calling ```TryAdd```
- get and use the value from ```s_nodeToMemberNames``` if TryAdd fails.

Reply:
Unfortunately, netstandard only has these members (so no 'TryAdd'):

![image](https://github.com/dotnet/roslyn/assets/4564579/c03cacf8-28d1-4f43-885d-6849db53c44f)

i can do this with a TryGetValue (non-allocating), then in the rarer case do the allocating path that has the callback.  The lock form has the benefit of never allocating, but can be avoided if we're ok with the alloc here.

Reply:
> Unfortunately, netstandard only has these members (so no

Then, I think, we should simply use ```GetValue``` method and return what it returns. And get rid of the conditional compilation. If closure is a big concern, we could use ```PooledDelegates``` to mitigate that.

Apart from this. Here are the issues I see with the code in commit 68:
- The NET7_0 branch doesn't reuse cached value when it loses the race. The reuse of the cached value was specifically suggested in the initial comment.
- The NET6_0 branch, the upfront ```TryGetValue``` check doesn't add any guarantees. We can still lose, and then we overwrite the value placed by the winner in the cache. Avoiding the overwrite was specifically suggested in the initial comment.
- The ```#else``` branch, the upfront ```TryGetValue``` check doesn't add any guarantees. We can still lose. The value returned by ```GetValue``` is ignored, meaning we do not  reuse cached value when we lose the race. The reuse of the cached value was specifically suggested in the initial comment.
- The NET6_0 and ```#else``` branches, we are ignoring the value returned by ```TryGetValue```, i.e. not reusing the value from the cache. The reuse of the cached value was specifically suggested in the initial comment.

Reply:
Switched to pooled delegate, and removed conditional branches.  My preference is to write idiomatic code once the apis are available, but i'm ok with an approach that is suboptimal on some runtimes, but is consistent across all our targs.

WRT to the feedback given above: 

> The NET7_0 branch doesn't reuse cached value when it loses the race.

That's not the case.  All branches ultimately went back and retrieved the value from the CWT after their attempt to add.  That way if it lost, it would still pick up the value it lost to.

> The NET6_0 branch, the upfront TryGetValue check doesn't add any guarantees. We can still lose, and then we overwrite the value placed by the winner in the cache. Avoiding the overwrite was specifically suggested in the initial comment.

Correct.  But it reduced a potential for always overwriting to one where it would only overwrite in the presence of an extremely narrow race.  I made the call that that was entirely fine as it woudl be unlikely to ever get hit in practice.  And in the pathological case where it did, the only downside was not reusing memory.  Given that the *Existing* behavior of the decl tree is to not reuse memory, this was simply no worse than before, and only for a narrow race.

> The #else branch, the upfront TryGetValue check doesn't add any guarantees. We can still lose. The value returned by GetValue is ignored, meaning we do not reuse cached value when we lose the race. 

Same as above.  And the reason for ignoring the GetValue was intentional as we always unilaterally read back from the cache to defer to the winning thread.

 > The NET6_0 and #else branches, we are ignoring the value returned by TryGetValue, i.e. not reusing the value from the cache. The reuse of the cached value was specifically suggested in the initial comment.

As per the above, we always attempt to reread a value stored in the cache from another thread, and we attempted to avoid allocations where possible.  In the event of a rare race, some of these branches might overwrite, but that would be no worse than the current situation.